### ISSUE TITLE
[Bug]: fused_moe_kernel compile bug

### ISSUE THREAD

**Issue by jeejeelee** [vllm-project/vllm#6103]
### Your current environment

```text
Collecting environment information...
PyTorch version: 2.3.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.0
Libc version: glibc-2.31

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-107-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A800 80GB PCIe
GPU 1: NVIDIA A800 80GB PCIe
GPU 2: NVIDIA A800 80GB PCIe
GPU 3: NVIDIA A800 80GB PCIe
GPU 4: NVIDIA A800 80GB PCIe
GPU 5: NVIDIA A800 80GB PCIe
GPU 6: NVIDIA A800 80GB PCIe
GPU 7: NVIDIA A800 80GB PCIe

Nvidia driver version: 550.54.15
cuDNN version: Probably one of the following:
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn.so.8.9.6
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.9.6
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.9.6
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.9.6
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.9.6
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.9.6
/usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 57 bits virtual
CPU(s):                             112
On-line CPU(s) list:                0-111
Thread(s) per core:                 2
Core(s) per socket:                 28
Socket(s):                          2
NUMA node(s):                       2
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              106
Model name:                         Intel(R) Xeon(R) Gold 6330 CPU @ 2.00GHz
Stepping:                           6
CPU MHz:                            2000.000
CPU max MHz:                        3100.0000
CPU min MHz:                        800.0000
BogoMIPS:                           4000.00
Virtualization:                     VT-x
L1d cache:                          2.6 MiB
L1i cache:                          1.8 MiB
L2 cache:                           70 MiB
L3 cache:                           84 MiB
NUMA node0 CPU(s):                  0-27,56-83
NUMA node1 CPU(s):                  28-55,84-111
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI Syscall hardening, KVM SW loop
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] mypy==1.9.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] nvidia-nccl-cu12==2.20.5
[pip3] torch==2.3.0
[pip3] torchvision==0.18.0
[pip3] transformers==4.42.1
[pip3] triton==2.3.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] nvidia-nccl-cu12          2.20.5                   pypi_0    pypi
[conda] torch                     2.3.0                    pypi_0    pypi
[conda] torchvision               0.18.0                   pypi_0    pypi
[conda] transformers              4.42.1                   pypi_0    pypi
[conda] triton                    2.3.0                    pypi_0    pypi
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.5.0.post1
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
[4mGPU0	GPU1	GPU2	GPU3	GPU4	GPU5	GPU6	GPU7	CPU Affinity	NUMA Affinity	GPU NUMA ID[0m
GPU0	 X 	NV8	PXB	PXB	PXB	PXB	PXB	PXB	0-27,56-83	0		N/A
GPU1	NV8	 X 	PXB	PXB	PXB	PXB	PXB	PXB	0-27,56-83	0		N/A
GPU2	PXB	PXB	 X 	PXB	PXB	PXB	PXB	NV8	0-27,56-83	0		N/A
GPU3	PXB	PXB	PXB	 X 	NV8	PXB	PXB	PXB	0-27,56-83	0		N/A
GPU4	PXB	PXB	PXB	NV8	 X 	PXB	PXB	PXB	0-27,56-83	0		N/A
GPU5	PXB	PXB	PXB	PXB	PXB	 X 	NV8	PXB	0-27,56-83	0		N/A
GPU6	PXB	PXB	PXB	PXB	PXB	NV8	 X 	PXB	0-27,56-83	0		N/A
GPU7	PXB	PXB	NV8	PXB	PXB	PXB	PXB	 X 	0-27,56-83	0		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

```

## Description

#5036 will raise the following error when using TP>1. Similar issues also persisted when I tested the Qwen1.5-MoE model 
 **due to the use of the fused_moe_kernel  triton kernel**.

```shell
[rank0]:   File "/root/anaconda3/envs/py310_vllm/lib/python3.10/site-packages/triton/compiler/compiler.py", line 239, in __init__
[rank0]:     self.asm = {
[rank0]:   File "/root/anaconda3/envs/py310_vllm/lib/python3.10/site-packages/triton/compiler/compiler.py", line 240, in <dictcomp>
[rank0]:     file.suffix[1:]: file.read_bytes() if file.suffix[1:] == driver.binary_ext else file.read_text()
[rank0]:   File "/root/anaconda3/envs/py310_vllm/lib/python3.10/pathlib.py", line 1134, in read_text
[rank0]:     with self.open(mode='r', encoding=encoding, errors=errors) as f:
[rank0]:   File "/root/anaconda3/envs/py310_vllm/lib/python3.10/pathlib.py", line 1119, in open
[rank0]:     return self._accessor.open(self, mode, buffering, encoding, errors,
[rank0]: FileNotFoundError: [Errno 2] No such file or directory: '/root/.triton/cache/76c543c0d8742e904f283b49e0ee704b/_sgmv_expand_kernel.json.tmp.pid_450752_42384ec9-f78e-40b0-b440-c41643b19290'
```

## Possible solutions
I believe it's a triton bug. I then debugged the triton code and investigated the related issue. and found that the cause of this issue is well explained [here](https://github.com/IBM/vllm/pull/35#issuecomment-2145591548).  Until triton officially addresses this issue, we may need to resolve it in a similar manner. Another approach is to set `distributed_executor_backend` to `ray ` when using the triton kernel.

## Reproducible Code

```python
from vllm import LLM, SamplingParams

# Sample prompts.
prompts = [
    "Hello, my name is",
    "The president of the United States is",
    "The capital of France is",
    "The future of AI is",
]
# Create a sampling params object.
sampling_params = SamplingParams(
    temperature=0.8,
    top_p=0.95,
    logprobs=1,
    prompt_logprobs=1,
    max_tokens=128,
)

# Create an LLM.
model_path = "Qwen1.5-MoE-A2.7B-Chat"  # Other models using the moe triton operator should also work.

llm = LLM(
    model=model_path,
    # enable_lora=True,
    trust_remote_code=True,
    gpu_memory_utilization=0.4,
    tensor_parallel_size=8,
    enforce_eager=False

)
# Generate texts from the prompts. The output is a list of RequestOutput objects
# that contain the prompt, generated text, and other information.
outputs = llm.generate(prompts, sampling_params)
# Print the outputs.
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")

```

**Comment by jeejeelee** [comment]
cc @njhill

**Comment by tdoublep** [comment]
Yeah, we've fixed this issue on our fork (as you found [here](https://github.com/IBM/vllm/pull/35)). Let me create a PR to contribute the fix upstream.

**Comment by randxie** [comment]
I was able to work around this bug by building the latest Triton code from source.

**Comment by jeejeelee** [comment]
> I was able to work around this bug by building the latest Triton code from source.

It seems that the triton main has updated the related code, I don't have more time to dive into this update. For now, I believe #6140 can serve as a temporary solution to address this issue. Once VLLM updates the triton version, we can revisit this issue again.

**Comment by tdoublep** [comment]
@randxie Interesting. I actually tried to test [these changes ](https://github.com/triton-lang/triton/pull/3544) that were merged into Triton main in our fork, but it didn't help. I wonder what else has changed in the meantime. 

Agree that getting the fix from Triton is the long-term solution though.

**Comment by randxie** [comment]
@tdoublep I believe there were more changes involved. If you check the log above, Triton tried to load files with naming pattern as `*.tmp.pid_*`, but when you go into the container, the file is not generated. I haven't spent time on figuring out the exact change, just want to provide a data point that latest Triton has fixed this issue.

**Comment by LSC527** [comment]
removing the entire triton cache dir every time before run vllm can be a temp workaround, when I run deepseekv2.

**Comment by tdoublep** [comment]
There was a PR merged into Triton yesterday that tries to address this issue: https://github.com/triton-lang/triton/pull/4295

**Comment by tdoublep** [comment]
I have verified that (a) the fix from https://github.com/triton-lang/triton/pull/4295 is included in the Triton v3.0.0 release and (b) that using `triton==3.0.0` fixes the issue.

**Comment by tdoublep** [comment]
Hmm I tried to bump Triton version but get conflict with torch dependencies:
```
INFO: pip is looking at multiple versions of torch to determine which version is compatible with other requirements. This could take a while.
ERROR: Cannot install -r requirements-cuda.txt (line 7) and triton>=3.0.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested triton>=3.0.0
    torch 2.3.0 depends on triton==2.3.0; platform_system == "Linux" and platform_machine == "x86_64" and python_version < "3.12"

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict
```
IDK how long we will need to wait until torch moves to `triton==3.0.0`. 

It might make sense to merge #6140 in the meantime.

**Comment by tdoublep** [comment]
Actually, it looks like torch have `triton==3.0.0` pinned on their main branch: https://github.com/pytorch/pytorch/blob/main/.ci/docker/triton_version.txt
https://github.com/pytorch/pytorch/blob/cb2bce98de5bba1c5415a08d7b17c608f6c77b62/.circleci/scripts/binary_populate_env.sh#L81

What is weird though is that this change was made 6 months ago, but triton 3.0.0 was only released yesterday? idgi

**Comment by tdoublep** [comment]
So I've been digging into this a bit more and here is a summary of my findings:

- Triton recently released v3.0.0, but it does **not** seem to include the [fix](https://github.com/triton-lang/triton/pull/4295). 
- Nevertheless, multiple people have [reported](https://github.com/vllm-project/vllm/issues/6103#issuecomment-2209298536), before yesterday, that recent Triton nightlies resolve the issue (or similar issues). 
- I can't really understand why any Triton nightly before that [PR](https://github.com/triton-lang/triton/pull/4295) would actually be safe w.r.t this issue (see the nice [diagram](https://private-user-images.githubusercontent.com/43726198/347152346-ffc20e0c-0404-4e7a-bd6c-022e710e97b9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA2ODc3MzksIm5iZiI6MTcyMDY4NzQzOSwicGF0aCI6Ii80MzcyNjE5OC8zNDcxNTIzNDYtZmZjMjBlMGMtMDQwNC00ZTdhLWJkNmMtMDIyZTcxMGU5N2I5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzExVDA4NDM1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhMDI3NjQ3NmNjY2EyZTNiNGE0YWYxMmNlMTU5OWI2YjFiNGM1NDM0NzYzZDdhNmUwYTAyODcwMDI1ZGZkZjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.119blHQluXxE_Qkv18sZCLtbu7YsuE2U4c3sfx_cuII) from the PR to understand why). 
- Therefore, in my view to be fully protected from this issue we will need to pull in whatever Triton version comes *after* 3.0.0, and this requires waiting for torch to upgrade to that and all other dependencies etc.
- I would therefore propose that we proceed to merge #6140 in the meantime.  
- Main remaining question is whether this issue can potentially also affect Ray deployments.

**Comment by jeejeelee** [comment]
> So I've been digging into this a bit more and here is a summary of my findings:
> 
> * Triton recently released v3.0.0, but it does **not** seem to include the [fix](https://github.com/triton-lang/triton/pull/4295) for this issue.
> * Nevertheless, multiple people have [reported](https://github.com/vllm-project/vllm/issues/6103#issuecomment-2209298536), before yesterday, that recent Triton nightlies resolve the issue (or similar issues).
> * I can't really understand why any Triton nightly before that [PR](https://github.com/triton-lang/triton/pull/4295) would actually be safe w.r.t this issue (see the nice [diagram](https://private-user-images.githubusercontent.com/43726198/347152346-ffc20e0c-0404-4e7a-bd6c-022e710e97b9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA2ODc3MzksIm5iZiI6MTcyMDY4NzQzOSwicGF0aCI6Ii80MzcyNjE5OC8zNDcxNTIzNDYtZmZjMjBlMGMtMDQwNC00ZTdhLWJkNmMtMDIyZTcxMGU5N2I5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzExVDA4NDM1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhMDI3NjQ3NmNjY2EyZTNiNGE0YWYxMmNlMTU5OWI2YjFiNGM1NDM0NzYzZDdhNmUwYTAyODcwMDI1ZGZkZjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.119blHQluXxE_Qkv18sZCLtbu7YsuE2U4c3sfx_cuII) from the PR to understand why). Everything is confounded by the fact this is a race condition and not deterministically reproducible.
> * Therefore, in my view to be fully protected from this issue we will need to pull in whatever Triton version comes _after_ 3.0.0, and this requires waiting for torch to upgrade to that and all other dependencies etc.
> * I would therefore propose that we proceed to merge [[Bugfix] Add custom Triton cache manager to resolve MoE MP issue  #6140](https://github.com/vllm-project/vllm/pull/6140) in the meantime.
> * Main remaining question is whether this issue can potentially also affect Ray deployments.

Firtly, I strongly agree with merging #6140. 

For `multiple people have https://github.com/vllm-project/vllm/issues/6103#issuecomment-2209298536, before yesterday, that recent Triton nightlies resolve the issue (or similar issues).`
This bug is occasional, in  my experience, with respect to the MOE model, the occurrence probability is higher when TP=8, typically during the `cuda graph` capture stage.  IMHO,  I think  it's possible that the error hasn't been reproduced again rather than being addressed

**Comment by solesensei** [comment]
Any updates on this?

**Comment by tdoublep** [comment]
Fix #6140 is ready from my pov, will try to get it approved and merged asap.
